genie-agent/src/main/java/com/netflix/genie/agent/execution/statemachine/actions/CleanupJobAction.java

```diff
@@ -44,7 +46,27 @@
     @Override
     protected Events executeStateAction(final ExecutionContext executionContext) {
         log.info("Cleaning up job...");
-        //TODO implement this action
+
+        final Deque<StateAction> deque = executionContext.getCleanupActions();
+        while (!deque.isEmpty()) {
+            final StateAction nextAction = deque.pop();
+
+            if (nextAction == this) {
+                // Skip self
+                continue;
```

Comment:
Is it even possible that there's a state after the current one? Should we break instead of continue?

Reply:
I wanted to avoid this state having knowledge of the rest of the machine topology.
It will execute cleanup for whatever is in the queue.
It knows that it can skip it's own cleanup, that's its prerogative.

Reply:
Makes sense I guess if you think cleanup could ever be called twice. No big deal I guess to keep it naive